src/routes/posts.ts

```diff
@@ -0,0 +1,125 @@
+import * as uuid from 'uuid';
+import * as Joi from '@hapi/joi';
+import { ContainerTypes, ValidatedRequest } from 'express-joi-validation';
+import * as pgsql from '../utils/pgsql';
+import { Post } from '../entities/post';
+import { POSTS_TABLE, USER_HEADER } from '../constants';
+import { BaseRequestSchema } from '../utils/schema';
+import { buildServiceError } from '../utils/error-builder';
+import { Errors } from '../errors';
+
+export const createPostSchema = {
+	body: Joi.object({
+		title: Joi.string().required(),
+	}),
+};
+
+export interface CreatePostRequestSchema extends BaseRequestSchema {
+	[ContainerTypes.Body]: {
+		title: string;
+	};
+}
+
+export async function create(req: ValidatedRequest<CreatePostRequestSchema>) {
+	const newPost = {
+		id: uuid.v4(),
+		user: req.headers[USER_HEADER],
+		createdAt: new Date(),
+		title: req.body.title,
+		votes: 0,
+		upvotePercentage: 0,
+		downvotePercentage: 0,
+	};
+
+	await pgsql.tryCatch(undefined, async (conn) => {
+		await conn.table(POSTS_TABLE).insert(transform(newPost));
+	});
+
+	return newPost;
+}
+
+export async function getAll() {
+	const posts = await pgsql.tryCatch(undefined, async (conn) => {
+		const query = conn
+			.table(POSTS_TABLE)
+			.select('*')
+			.orderBy([
+				{ column: 'upvote_percentage', order: 'desc' },
+				{ column: 'downvote_percentage', order: 'asc' },
+				{ column: 'votes', order: 'desc' },
+				{ column: 'created_at', order: 'desc' },
+			]);
+
+		return (await query).map(reverseTransform);
+	});
+
+	return posts;
+}
```

Comment:
In the challenge we say:

> "_Your solution should consider that if you have a first post with 600 up votes and 400 down votes means that you have 60% of up votes and 40% of down votes, and if you have another post with 6 up votes and 4 down votes you have the same % of the previous post, 60% of up votes and 40% of down votes._
> 
> _Hint: Note that the score in % of the two posts are equal, but the real "values" are significantly different._"

How could you distinguish the 600/400 post from the 6/4 post? Both have 60% upvotes. Which one should appear first and how could you calculate their score?

Reply:
In that specific scenario I assumed the post with more votes would come first. As per my README:

> When the list posts endpoint is called, the posts are ordered according to the following rules (in order of priority):
> 1. Upvote percentage, descending
> 2. Downvote percentage, ascending
> 3. Total votes, descending
> 4. Date created, descending
> 
> This gives us the following:
> - Posts with larger upvote percentage appear higher on the list
> - Posts with the more total votes appear higher than posts with the less votes but the same upvote percentage
> - Post with no votes appear higher than posts with 100% downvotes
> - If all else fails posts are ordered from most recent to oldest

I opted to use these simple rules to order posts intead of a score due to the simple and static nature of the project. In a real world scenario (e.g. Hacker News or Reddit) a score would definitely be more appropriate since it would also have to accomodate some kind of time-based decay (i.e. popular but old posts may appear lower on the list).

Reply:
👍